chemper/smirksify.py

```diff
@@ -218,49 +346,95 @@ def remove_and(self, input_all_ands):
         if len(input_all_ands) == 0:
             return input_all_ands, False
 
-        # otherwise randomly remove an AND decorator
-        all_ands = copy.deepcopy(input_all_ands)
-        all_ands.remove(random.choice(all_ands))
+        if random.rand() > 0.5:
+            # otherwise randomly remove an AND decorator
+            all_ands = copy.deepcopy(input_all_ands)
+            all_ands.remove(random.choice(all_ands))
+        else:
+            all_ands = list()
         return all_ands, True
 
+    def _get_item_and_remove_options(self, env):
+        """
+        Parameters
+        ----------
+        env: ChemicalEnvironment
+
+        Returns
+        -------
+        item: an Atom or Bond from the Chemical Environment
+        dec_opts: list
+            0 = has OR decorators to remove
+            1 = has AND decorators to remove
+            2 = is an atom that could be removed
+        """
+        items = list()
+        odds = list()
+        for a_b in env.getAtoms() + env.getBonds():
+            count = len(a_b.getANDtypes())
+            for o in a_b.getORtypes():
+                if o[0] != '*':
+                    count += 1
+                count += len(o[1])
+
+            # a wild card, atom with no index should be considered ([*])
+            # should also be on this list so it can be removed
+            if isinstance(a_b, CE.Atom) and not isinstance(a_b, CE.Bond) and env.isUnindexed(a_b):
+                count += 1
+
+            items.append(a_b)
+            odds.append(count)
+
+        # normalize odds to weights
+        odds = np.array(odds)
+
+        # if the odds for changing all atoms and bonds is zero return None
+        if odds.sum() == 0:
+            return None, list()
+
+        weights = odds / odds.sum()
+
+        # choose an atom or bond with the probabilities:
+        item = random.choice(items, p=weights)
+        dec_opts = list()
+        if len(item.getORtypes()) > 0:
+            dec_opts.append('remove_ors')
+        if len(item.getANDtypes()) > 0:
+            dec_opts.append('remove_ands')
+
+        if not isinstance(item, CE.Bond): # then it is an atom
+            if env.getValence(item) == 1 and env.isUnindexed(item):
+                dec_opts.append('remove_atom')
+
+        return item, dec_opts
+
     def remove_decorator(self, smirks):
```

Comment:
Easiest to check if you start with this function and see if the logic in this function and its dependencies follow what is described in the PR's description. Because of all the changes it might be easier to switch the "Diff settings" to Split instead of unified.